core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpParser.java

```diff
@@ -0,0 +1,635 @@
+            return false;
+        }
+
+        errorCode = ErrorCode.INVALID_FIELD_SEPARATOR;
+        return false;
+    }
+
+    private boolean expectTableName(byte endOfEntityByte, long bufHi) {
+        tagsComplete = endOfEntityByte == (byte) ' ';
+        if (endOfEntityByte == (byte) ',' || tagsComplete) {
+            measurementName.of(entityLo, bufAt - nEscapedChars);
+            entityHandler = entityNameHandler;
+            return true;
+        }
+
+        if (entityLo == bufAt) {
+            errorCode = ErrorCode.EMPTY_LINE;
+        } else {
+            errorCode = ErrorCode.NO_FIELDS;
+        }
+        return false;
+    }
+
+    private boolean expectTimestamp(byte endOfEntityByte, long bufHi) {
+        try {
+            if (endOfEntityByte == (byte) '\n') {
+                timestamp = Numbers.parseLong(charSeq.of(entityLo, bufAt - nEscapedChars));
+                entityHandler = null;
+                return true;
+            }
+            errorCode = ErrorCode.INVALID_FIELD_SEPARATOR;
+            return false;
+        } catch (NumericException ex) {
+            errorCode = ErrorCode.INVALID_TIMESTAMP;
+            return false;
+        }
+    }
+
+    public enum ParseResult {
+        MEASUREMENT_COMPLETE, BUFFER_UNDERFLOW, ERROR
+    }
+
+    public enum ErrorCode {
+        EMPTY_LINE,
+        NO_FIELDS,
+        INCOMPLETE_TAG,
+        INCOMPLETE_FIELD,
+        INVALID_FIELD_SEPARATOR,
+        INVALID_TIMESTAMP,
+        INVALID_FIELD_VALUE,
+        INVALID_FIELD_VALUE_STR_UNDERFLOW,
+        NONE
+    }
+
+    private interface EntityHandler {
+        boolean completeEntity(byte endOfEntityByte, long bufHi);
+    }
+
+    public class ProtoEntity {
+        private final DirectByteCharSequence name = new DirectByteCharSequence();
+        private final DirectByteCharSequence value = new DirectByteCharSequence();
+        private byte type = ENTITY_TYPE_NONE;
+        private long integerValue;
+        private boolean booleanValue;
+        private double floatValue;
+        private long timestampValue;
+
+        public boolean getBooleanValue() {
+            return booleanValue;
+        }
+
+        public double getFloatValue() {
+            return floatValue;
+        }
+
+        public long getIntegerValue() {
+            return integerValue;
+        }
+
+        public long getTimestampValue() {
+            return timestampValue;
+        }
+
+        public DirectByteCharSequence getName() {
+            return name;
+        }
+
+        public byte getType() {
+            return type;
+        }
+
+        public DirectByteCharSequence getValue() {
+            return value;
+        }
+
+        public void shl(long shl) {
+            name.shl(shl);
+            value.shl(shl);
+        }
+
+        private void clear() {
+            type = ENTITY_TYPE_NONE;
+        }
+
+        private boolean parse(byte last, int valueLen) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.